File: src/Algorithm/Arrays/Pow(x,n)/pow.py
# Problem link: https://leetcode.com/problems/powx-n/

class Solution:
    def myPow(self, x: float, n: int) -> float:
        def recurMe(m,n):
            if n==0:
                return 1
            if n==1:
                return m
            else:
                temp=recurMe(m,n//2)
                if n%2!=0:
                    return temp*temp*m
                else:
                    return temp*temp
        if n<0:
            return (1/recurMe(x,abs(n)))
        return recurMe(x,abs(n))

File: src/Algorithm/Arrays/Pow(x,n)/pow.cpp
/*
    Problem Link: https://leetcode.com/problems/powx-n/
    Implement pow(x, n), which calculates x raised to the power n
    Ex. x = 2 n = 10 -> 1024, x = 2.1 n = 3 -> 9.261, x = 2 n = -2 -> 0.25

    Divide-and-conquer, even x^n = A * A, odd x^n = A * A * x

    Time: O(log n)
    Space: O(1) (if we use a global variable to store the result)
*/

class Solution {
    private:
    double power(double x, int n){
        if(n == 0){
            return 1;
        }
        double ans = power(x,n/2);
        if(n%2 == 0){
            return ans*ans;
        }
        else{
            return ans*ans*x;
        }
    }
public:
    double myPow(double x, int n) {
        if(n<0)
            return (1/power(x,n));
        else
            return  power(x,n);
    }
};